src/main/java/io/lettuce/core/StatefulRedisConnectionImpl.java

```diff
@@ -199,37 +184,30 @@ public <T> RedisCommand<K, V, T> dispatch(RedisCommand<K, V, T> command) {
 
         RedisCommand<K, V, T> toSend = preProcessCommand(command);
         RedisCommand<K, V, T> result = super.dispatch(toSend);
-        if (toSend.getType() == EXEC || toSend.getType() == DISCARD) {
-            inTransaction.set(false);
-            setCredentials(credentialsRef.getAndSet(null));
-        }
+        RedisCommand<K, V, T> finalCommand = postProcessCommand(result);
 
-        return result;
+        return finalCommand;
     }
 
     @Override
     public Collection<RedisCommand<K, V, ?>> dispatch(Collection<? extends RedisCommand<K, V, ?>> commands) {
 
         List<RedisCommand<K, V, ?>> sentCommands = new ArrayList<>(commands.size());
 
-        boolean transactionComplete = false;
-        for (RedisCommand<K, V, ?> o : commands) {
-            RedisCommand<K, V, ?> command = preProcessCommand(o);
-            sentCommands.add(command);
-            if (command.getType() == EXEC) {
-                transactionComplete = true;
-            }
-            if (command.getType() == MULTI || command.getType() == DISCARD) {
-                transactionComplete = false;
-            }
-        }
+        commands.forEach(o -> {
+            RedisCommand<K, V, ?> preprocessed = preProcessCommand(o);
+            sentCommands.add(preprocessed);
+        });
 
-        Collection<RedisCommand<K, V, ?>> result = super.dispatch(sentCommands);
-        if (transactionComplete) {
-            inTransaction.set(false);
-            setCredentials(credentialsRef.getAndSet(null));
-        }
-        return result;
+        super.dispatch(sentCommands);
+
+        sentCommands.forEach(this::postProcessCommand);
+        return sentCommands;
```

Comment:
Can you give an example of when this would not work?

If an `AUTH` command arrives before we pre-process a command in the batch then we still have not started the transaction. It is generally the same as with processing individual commands in the straight case (imagine we have 5 threads that are sending commands and one of the commands is a `MULTI`.

Reply:
Can you give an example of when this would not work?

If an `AUTH` command arrives before we pre-process a command in the batch then we still have not started the transaction. It is generally the same as with processing individual commands in the straight case (imagine we have 5 threads that are sending commands and one of the commands is a `MULTI`.

Reply:
We have just dispatched the whole batch of commands  (  MULTI , DISCARD , MULTI)
we will emit auth command just after we post-process DISCARD,  when we should not, since the end result is that we are currently in transaction